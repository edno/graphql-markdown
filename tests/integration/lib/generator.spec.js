const mock = require("mock-fs");
const path = require("path"); // to be loaded after mock-fs
const dirTree = require("directory-tree");

const generateDocFromSchema = require("@/lib/generator");

const EXPECT_PATH = path.join(
  __dirname,
  "__expect__",
  __OS__,
  path.basename(__filename),
);

describe("lib", () => {
  beforeEach(() => {
    mock({
      node_modules: mock.load(path.resolve(__dirname, "../../../node_modules")),
      __data__: mock.load(path.resolve(__dirname, "../../__data__"), {
        lazy: false,
      }),
      output: {},
      assets: {
        "generated.md": "Dummy homepage for tweet.graphql",
        "sidebar.json": mock.load(require.resolve("@assets/sidebar.json"), {
          lazy: false,
        }),
      },
      tmp: {},
    });
  });

  afterEach(() => {
    mock.restore();
  });

  describe("renderer", () => {
    describe("generateDocFromSchema()", () => {
      test("generates Markdown document structure from GraphQL schema", async () => {
        expect.assertions(2);

        await generateDocFromSchema(
          "graphql",
          "__data__/tweet.graphql",
          "output",
          "docs",
          "assets/generated.md",
          "SCHEMA-DIFF",
          "tmp",
        );

        const outputFolder = dirTree("output");
        const tmpFolder = dirTree("tmp");

        mock.restore(); // see https://github.com/tschaub/mock-fs#caveats

        expect(JSON.stringify(outputFolder, null, 2)).toMatchFile(
          path.join(EXPECT_PATH, `generateDocFromSchemaOutputFolder.hash`),
        );
        expect(JSON.stringify(tmpFolder, null, 2)).toMatchFile(
          path.join(EXPECT_PATH, `generateDocFromSchemaTmpFolder.hash`),
        );
      });
    });
  });
});
