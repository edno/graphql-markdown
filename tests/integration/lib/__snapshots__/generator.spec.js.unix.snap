// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/directives/_category_.yml",
          "size": 20,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "deprecated.mdx",
          "path": "output/directives/deprecated.mdx",
          "size": 470,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "include.mdx",
          "path": "output/directives/include.mdx",
          "size": 273,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "skip.mdx",
          "path": "output/directives/skip.mdx",
          "size": 255,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "specified-by.mdx",
          "path": "output/directives/specified-by.mdx",
          "size": 285,
          "type": "file",
        },
      ],
      "name": "directives",
      "path": "output/directives",
      "size": 1303,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/enums/_category_.yml",
          "size": 15,
          "type": "file",
        },
      ],
      "name": "enums",
      "path": "output/enums",
      "size": 15,
      "type": "directory",
    },
    Object {
      "extension": ".md",
      "name": "generated.md",
      "path": "output/generated.md",
      "size": 32,
      "type": "file",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/inputs/_category_.yml",
          "size": 16,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "tweet-content.mdx",
          "path": "output/inputs/tweet-content.mdx",
          "size": 258,
          "type": "file",
        },
      ],
      "name": "inputs",
      "path": "output/inputs",
      "size": 274,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/interfaces/_category_.yml",
          "size": 20,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "node.mdx",
          "path": "output/interfaces/node.mdx",
          "size": 148,
          "type": "file",
        },
      ],
      "name": "interfaces",
      "path": "output/interfaces",
      "size": 168,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/mutations/_category_.yml",
          "size": 19,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "create-tweet.mdx",
          "path": "output/mutations/create-tweet.mdx",
          "size": 352,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "delete-tweet.mdx",
          "path": "output/mutations/delete-tweet.mdx",
          "size": 224,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "mark-tweet-read.mdx",
          "path": "output/mutations/mark-tweet-read.mdx",
          "size": 294,
          "type": "file",
        },
      ],
      "name": "mutations",
      "path": "output/mutations",
      "size": 889,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/objects/_category_.yml",
          "size": 17,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "media.mdx",
          "path": "output/objects/media.mdx",
          "size": 212,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "meta.mdx",
          "path": "output/objects/meta.mdx",
          "size": 150,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "notification.mdx",
          "path": "output/objects/notification.mdx",
          "size": 299,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "stat.mdx",
          "path": "output/objects/stat.mdx",
          "size": 496,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "tweet.mdx",
          "path": "output/objects/tweet.mdx",
          "size": 633,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "user.mdx",
          "path": "output/objects/user.mdx",
          "size": 924,
          "type": "file",
        },
      ],
      "name": "objects",
      "path": "output/objects",
      "size": 2731,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/queries/_category_.yml",
          "size": 17,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "notifications-meta.mdx",
          "path": "output/queries/notifications-meta.mdx",
          "size": 240,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "tweet.mdx",
          "path": "output/queries/tweet.mdx",
          "size": 205,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "tweets-meta.mdx",
          "path": "output/queries/tweets-meta.mdx",
          "size": 219,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "tweets.mdx",
          "path": "output/queries/tweets.mdx",
          "size": 455,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "user.mdx",
          "path": "output/queries/user.mdx",
          "size": 267,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "users.mdx",
          "path": "output/queries/users.mdx",
          "size": 211,
          "type": "file",
        },
      ],
      "name": "queries",
      "path": "output/queries",
      "size": 1614,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/scalars/_category_.yml",
          "size": 17,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "boolean.mdx",
          "path": "output/scalars/boolean.mdx",
          "size": 123,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "date.mdx",
          "path": "output/scalars/date.mdx",
          "size": 73,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "id.mdx",
          "path": "output/scalars/id.mdx",
          "size": 385,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "int.mdx",
          "path": "output/scalars/int.mdx",
          "size": 187,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "sri.mdx",
          "path": "output/scalars/sri.mdx",
          "size": 315,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "string.mdx",
          "path": "output/scalars/string.mdx",
          "size": 246,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "url.mdx",
          "path": "output/scalars/url.mdx",
          "size": 70,
          "type": "file",
        },
      ],
      "name": "scalars",
      "path": "output/scalars",
      "size": 1416,
      "type": "directory",
    },
    Object {
      "extension": ".js",
      "name": "sidebar-schema.js",
      "path": "output/sidebar-schema.js",
      "size": 114,
      "type": "file",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/subscriptions/_category_.yml",
          "size": 23,
          "type": "file",
        },
        Object {
          "extension": ".mdx",
          "name": "notifications.mdx",
          "path": "output/subscriptions/notifications.mdx",
          "size": 260,
          "type": "file",
        },
      ],
      "name": "subscriptions",
      "path": "output/subscriptions",
      "size": 283,
      "type": "directory",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".yml",
          "name": "_category_.yml",
          "path": "output/unions/_category_.yml",
          "size": 16,
          "type": "file",
        },
      ],
      "name": "unions",
      "path": "output/unions",
      "size": 16,
      "type": "directory",
    },
  ],
  "name": "output",
  "path": "output",
  "size": 8855,
  "type": "directory",
}
`;

exports[`lib renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 2`] = `
Object {
  "children": Array [
    Object {
      "extension": "",
      "name": ".schema",
      "path": "tmp/.schema",
      "size": 64,
      "type": "file",
    },
    Object {
      "extension": ".graphql",
      "name": "schema.graphql",
      "path": "tmp/schema.graphql",
      "size": 1351,
      "type": "file",
    },
  ],
  "name": "tmp",
  "path": "tmp",
  "size": 1415,
  "type": "directory",
}
`;
